src/bp/nlu-core/training-worker-queue.ts

```diff
@@ -0,0 +1,253 @@
+      const newWorker = await this._createNewWorker()
+      this.waitingWorkers.push(newWorker)
+    }
+
+    const worker = this.waitingWorkers.pop()!
+    this.activeWorkers[trainSessionId] = worker
+
+    let output: TrainOutput
+    try {
+      output = await this._startTraining(worker, input, progress)
+    } catch (err) {
+      const isTrainingCanceled = err instanceof TrainingCanceledError
+      if (!isTrainingCanceled) {
+        this._prepareForNextTraining(trainSessionId)
+      }
+      throw err
+    }
+    this._prepareForNextTraining(trainSessionId)
+    return output
+  }
+
+  private _prepareForNextTraining(trainSessionId: string) {
+    const worker = this.activeWorkers[trainSessionId]
+    this.waitingWorkers.unshift(worker)
+    delete this.activeWorkers[trainSessionId]
+  }
+
+  private async _startTraining(
+    workerId: number,
+    input: TrainInput,
+    progress: (x: number) => void
+  ): Promise<TrainOutput> {
+    const msg: OutgoingMessage = { type: 'start_training', destWid: workerId, payload: { input } }
+
+    return new Promise((resolve, reject) => {
+      const handler = (msg: IncomingMessage) => {
+        if (msg.type === 'training_done') {
+          process.off('message', handler)
+          resolve(msg.payload.output!)
+        }
+        if (msg.type === 'training_error') {
+          process.off('message', handler)
+          reject(new Error(msg.payload.error!))
+        }
+        if (msg.type === 'training_canceled' && msg.srcWid === workerId) {
+          process.off('message', handler)
+          reject(new TrainingCanceledError())
+        }
+        if (msg.type === 'training_progress') {
+          progress(msg.payload.progress!)
+        }
+      }
+      process.send!(msg)
+      process.on('message', handler)
+    })
+  }
+
+  private _createNewWorker(): Promise<number> {
+    const { config } = this
+    const msg: OutgoingMessage = { type: 'make_new_worker', payload: { config } }
+
+    return new Promise(resolve => {
+      const handler = (msg: IncomingMessage) => {
+        if (msg.type === 'log') {
+          this._logMessage(msg)
+        }
+
+        if (msg.type === 'worker_ready') {
+          process.off('message', handler)
+          resolve(msg.srcWid)
+        }
+      }
+      process.send!(msg)
+      process.on('message', handler)
+    })
+  }
+
+  private _logMessage(msg: IncomingMessage) {
+    const log: Log = msg.payload.log!
+    log.info && this.logger.info(log.info)
+    log.warning && this.logger.warning(log.warning)
+    log.error && this.logger.error(log.error)
+  }
+}
+
+if (cluster.isMaster) {
+  function sendToWebWorker(msg: IncomingMessage) {
+    const webWorker = cluster.workers[process.WEB_WORKER]
+    webWorker?.isConnected() && webWorker.send(msg)
+  }
+
+  function sendToTrainingWorker(msg: OutgoingMessage) {
+    const worker = cluster.workers[msg.destWid!]
+    worker?.send(msg) // TODO: find out why this is sometimes undefined.
```

Comment:
what's undefined ? worker ? or .send ? did you stumble on that situation ?  is it because it got killed and not removed yet ? do we have a race condition ?

Reply:
discussed in person.